### DISCUSSION TITLE
Obtain codepoint for character

### DISCUSSION THREAD

**Discussion by wezm** [gleam-lang/gleam#1877] [Questions & support]
For Advent of Code day 3 there was this detail:

> Lowercase item types a through z have priorities 1 through 26.
> Uppercase item types A through Z have priorities 27 through 52.

Given a string containing only a-z, A-Z the idea was to calculate the "priority" of each character. 

`string.to_graphemes` appears to be the way to get the "characters" of a string. So I then had `List(String)` and wanted to get the code point of the character/grapheme that the String was holding. I couldn't work out how to do this. It seems like bit strings might be able to do it, this is what I tried:

```gleam
fn priority(item: String) -> Int {
  let cp = case <<item:utf8>> {
    <<a:utf8_codepoint>> -> a
  }
  // convert cp to Int here
}
```

but I couldn't work out how to convert `UtfCodepoint` into an `Int`. I note that the string module has a function to go the other way though.

Is this possible? In the end I just wrote out a big case expression:

```gleam
fn priority(item: String) -> Int {
  case item {
    "a" -> 1
    "b" -> 2
    "c" -> 3
    "d" -> 4
    "e" -> 5
    "f" -> 6
    "g" -> 7
    "h" -> 8
    "i" -> 9
    "j" -> 10
    "k" -> 11
    "l" -> 12
    "m" -> 13
    "n" -> 14
    "o" -> 15
    "p" -> 16
    "q" -> 17
    "r" -> 18
    "s" -> 19
    "t" -> 20
    "u" -> 21
    "v" -> 22
    "w" -> 23
    "x" -> 24
    "y" -> 25
    "z" -> 26
    "A" -> 27
    "B" -> 28
    "C" -> 29
    "D" -> 30
    "E" -> 31
    "F" -> 32
    "G" -> 33
    "H" -> 34
    "I" -> 35
    "J" -> 36
    "K" -> 37
    "L" -> 38
    "M" -> 39
    "N" -> 40
    "O" -> 41
    "P" -> 42
    "Q" -> 43
    "R" -> 44
    "S" -> 45
    "T" -> 46
    "U" -> 47
    "V" -> 48
    "W" -> 49
    "X" -> 50
    "Y" -> 51
    "Z" -> 52
  }
}
```

**Comment by inoas** [comment]
This should help you, https://github.com/gleam-lang/stdlib/pull/37 right, if it does you could thumb it :D

**Reply by wezm** [reply]
No sorry I don't think that helps. I need to be able to convert characters into codepoints but it looks like this is for trimming strings?

**Reply by inoas** [reply]
Sorry not `#37` but https://github.com/gleam-lang/stdlib/pull/374 - link was cut off by 1 char.

**Reply by wezm** [reply]
Right that's in the right space but `UtfCodepoint` still seems like a mostly opaque type. It can be constructed from an `Int` but I can't see a way to turn it back into an `Int`. Is there a way to do that?

**Reply by lpil** [reply]
A matching function to do this would be accepted 👍

**Reply by inoas** [reply]
https://github.com/gleam-lang/stdlib/pull/374/files also has a function that turns a string into a list of codepoint ints

**Comment by wezm** [comment]
Related to this query, is there a way to match on numeric ranges at the moment? My searching suggests that there isn't a pattern syntax for this but I thought I'd double check.

**Comment by hayleigh-dot-dev** [comment]
If you want to treat the string/grapheme as an int, that's all you need to do:
```
fn priority(item: String) -> Int {
  let <<codepoint:int>> = <<item:utf8>>
  // codepoint is an `Int` here
}
```

**Reply by inoas** [reply]
AFAIU crashes if `item` equals no codepoints (`""`) or more than 1 codepoint.
Did I understand correctly that in future that pattern match might not compile and/or be converted into a `Result`?
